File: tests/test_feature_detection.py
# -*- coding: utf-8 -*-
"""
Tests for feature_detection.py
"""

import pytest
from feature_detection import FeatureDetector, get_feature_detector, is_feature_available


class TestFeatureDetector:
    """Test feature detection system."""

    def test_initialization(self):
        """Test detector initializes without error."""
        detector = FeatureDetector()
        assert detector is not None
        assert detector._cache is not None

    def test_get_global_instance(self):
        """Test global instance is singleton."""
        detector1 = get_feature_detector()
        detector2 = get_feature_detector()
        assert detector1 is detector2

    def test_is_feature_available(self):
        """Test feature availability checks."""
        detector = get_feature_detector()

        # Check known features exist in cache
        assert "voice_stt" in detector._cache
        assert "voice_tts" in detector._cache
        assert "pdf" in detector._cache
        assert "rag" in detector._cache
        assert "audio" in detector._cache

    def test_get_status(self):
        """Test detailed status retrieval."""
        detector = get_feature_detector()
        status = detector.get_status("pdf")

        assert status is not None
        assert hasattr(status, "available")
        assert hasattr(status, "module_name")
        assert isinstance(status.available, bool)

    def test_get_unavailable_reason(self):
        """Test getting reason for unavailable features."""
        detector = get_feature_detector()

        # Get all unavailable features
        unavailable = detector.get_all_unavailable()

        for feature_name, status in unavailable.items():
            reason = detector.get_unavailable_reason(feature_name)
            assert isinstance(reason, str)
            assert len(reason) > 0

    def test_get_all_unavailable(self):
        """Test getting all unavailable features."""
        detector = get_feature_detector()
        unavailable = detector.get_all_unavailable()

        assert isinstance(unavailable, dict)
        # All values should have available=False
        for status in unavailable.values():
            assert status.available is False

    def test_unknown_feature(self):
        """Test handling of unknown feature."""
        detector = get_feature_detector()

        assert not is_feature_available("nonexistent_feature")
        assert detector.get_status("nonexistent_feature") is None


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: rust_output/src/test_feature_detection.rs
/// Tests for feature_detection::py

use anyhow::{Result, Context};
use crate::feature_detection::{FeatureDetector, get_feature_detector, is_feature_available};

/// Test feature detection system.
#[derive(Debug, Clone)]
pub struct TestFeatureDetector {
}

impl TestFeatureDetector {
    /// Test detector initializes without error.
    pub fn test_initialization(&self) -> () {
        // Test detector initializes without error.
        let mut detector = FeatureDetector();
        assert!(detector.is_some());
        assert!(detector._cache.is_some());
    }
    /// Test global instance is singleton.
    pub fn test_get_global_instance(&self) -> () {
        // Test global instance is singleton.
        let mut detector1 = get_feature_detector();
        let mut detector2 = get_feature_detector();
        assert!(detector1 == detector2);
    }
    /// Test feature availability checks.
    pub fn test_is_feature_available(&self) -> () {
        // Test feature availability checks.
        let mut detector = get_feature_detector();
        assert!(detector._cache.contains(&"voice_stt".to_string()));
        assert!(detector._cache.contains(&"voice_tts".to_string()));
        assert!(detector._cache.contains(&"pdf".to_string()));
        assert!(detector._cache.contains(&"rag".to_string()));
        assert!(detector._cache.contains(&"audio".to_string()));
    }
    /// Test detailed status retrieval.
    pub fn test_get_status(&self) -> () {
        // Test detailed status retrieval.
        let mut detector = get_feature_detector();
        let mut status = detector.get_status("pdf".to_string());
        assert!(status.is_some());
        assert!(/* hasattr(status, "available".to_string()) */ true);
        assert!(/* hasattr(status, "module_name".to_string()) */ true);
        assert!(/* /* isinstance(status.available, bool) */ */ true);
    }
    /// Test getting reason for unavailable features.
    pub fn test_get_unavailable_reason(&self) -> () {
        // Test getting reason for unavailable features.
        let mut detector = get_feature_detector();
        let mut unavailable = detector.get_all_unavailable();
        for (feature_name, status) in unavailable.iter().iter() {
            let mut reason = detector.get_unavailable_reason(feature_name);
            assert!(/* /* isinstance(reason, str) */ */ true);
            assert!(reason.len() > 0);
        }
    }
    /// Test getting all unavailable features.
    pub fn test_get_all_unavailable(&self) -> () {
        // Test getting all unavailable features.
        let mut detector = get_feature_detector();
        let mut unavailable = detector.get_all_unavailable();
        assert!(/* /* isinstance(unavailable, dict) */ */ true);
        for status in unavailable.values().iter() {
            assert!(status.available == false);
        }
    }
    /// Test handling of unknown feature.
    pub fn test_unknown_feature(&self) -> () {
        // Test handling of unknown feature.
        let mut detector = get_feature_detector();
        assert!(!is_feature_available("nonexistent_feature".to_string()));
        assert!(detector.get_status("nonexistent_feature".to_string()).is_none());
    }
}
